Core/Executor/ContentManager.php

```diff
@@ -595,33 +595,137 @@ public function generateMigration(array $matchCondition, $mode, array $context =
                 reset($field);
                 $fieldIdentifier = key($field);
                 $fieldValue = $field[$fieldIdentifier];
+
+                $convertedFields[$fieldIdentifier] = $fieldValue;
             } else {
-                $fieldIdentifier = $key;
-                $fieldValue = $field;
+                $convertedFields[$key] = $field;
             }
+            $i++;
+        }
 
-            if (!isset($contentType->fieldDefinitionsByIdentifier[$fieldIdentifier])) {
-                throw new \Exception("Field '$fieldIdentifier' is not present in content type '{$contentType->identifier}'");
+        return $convertedFields;
+    }
+
+    /**
+     * @param array $fields
+     * @param string $language
+     * @return array
+     */
+    protected function parseSingleLangFields(array $fields, $language)
+    {
+        $fieldsList = [];
+        foreach ($fields as $fieldIdentifier => $fieldValue) {
+            $fieldsList[] = [$fieldIdentifier, $fieldValue, $language];
+        }
+
+        return $fieldsList;
+    }
+
+    /**
+     * @param array $fields
+     * @return array
+     * @throws \Exception
+     */
+    protected function parseMultiLangFields(array$fields)
+    {
+        $fieldsList = [];
+
+        foreach ($fields as $key => $field) {
+            if (!is_array($field)) {
+                throw new \Exception("Field '$key' should use multilang syntax as other fields in this step");
             }
 
-            $fieldDefinition = $contentType->fieldDefinitionsByIdentifier[$fieldIdentifier];
-            $fieldValue = $this->getFieldValue($fieldValue, $fieldDefinition, $contentType->identifier, $step->context);
+            foreach($field as $languageCode => $value) {
+                $fieldsList[] = [$key, $value, $languageCode];
+            }
+        }
 
-            $createOrUpdateStruct->setField($fieldIdentifier, $fieldValue, $this->getLanguageCode($step));
+        return $fieldsList;
+    }
 
-            $i++;
+    /**
+     * Checks whatever at least one of field is using multilang syntax.
+     *
+     * @param array $fields
+     * @return bool
+     */
+    protected function hasLanguageCodesAsKeys(array $fields)
+    {
+        $languages = $this->getContentLanguages();
+        $languageCodes = array_combine($languages, $languages);
+        $hasLanguageCodesAsKeys = false;
+
+        foreach ($fields as $fieldIdentifier => $fieldData) {
+            if (!is_array($fieldData)) {
+                continue;
+            }
+
+            foreach ($fieldData as $key => $data) {
+                if (array_key_exists($key, $languageCodes)) {
+                    $hasLanguageCodesAsKeys = true;
+                }
+            }
         }
+
+        return $hasLanguageCodesAsKeys;
+    }
+
+    /**
+     * Returns all Languages.
+     *
+     * @return array
+     */
+    protected function getContentLanguages()
+    {
+        static $languages;
```

Comment:
I think that this (static) might not be 100% safe, as there is a contrived case where we could access many repos at the same time